# LAP: Aufgabenserie 08

## Aufgabe 1 - Fehlerbehandlung (Laufzeitfehler)
### a) Eigene Eingabe
Das folgende Programm soll so erweitert werden, dass ganze Zahlen als Zähler und Nenner eingegeben werden können.

In [21]:
from fractions import Fraction

print(Fraction(5, 6) + Fraction(1, 3))      # 5/6 + 1/3
print(Fraction(1, 3) - Fraction(2, 3))      # 1/3 - 2/3
print(Fraction(3, 4) * Fraction(2, 3))      # 3/4 * 2/3
print(Fraction(2, 7) / Fraction(2, 9))      # 2/7 / 2/9

7/6
-1/3
1/2
9/7


In [22]:
from fractions import Fraction

# Eingabe des ersten Bruchs
zaehler1 = int(input("Gib den Zähler des ersten Bruchs ein: "))
nenner1 = int(input("Gib den Nenner des ersten Bruchs ein: "))

# Eingabe des zweiten Bruchs
zaehler2 = int(input("Gib den Zähler des zweiten Bruchs ein: "))
nenner2 = int(input("Gib den Nenner des zweiten Bruchs ein: "))
    
# Erstellen der Brüche
bruch1 = Fraction(zaehler1, nenner1)
bruch2 = Fraction(zaehler2, nenner2)
    
# Berechnungen
print(f"{bruch1} + {bruch2} = {bruch1 + bruch2}")
print(f"{bruch1} - {bruch2} = {bruch1 - bruch2}")
print(f"{bruch1} * {bruch2} = {bruch1 * bruch2}")
print(f"{bruch1} / {bruch2} = {bruch1 / bruch2}")

1/5 + 2/3 = 13/15
1/5 - 2/3 = -7/15
1/5 * 2/3 = 2/15
1/5 / 2/3 = 3/10


### b) Fehlerbehandlung (Laufzeitfehler)
Jetzt soll das Programm so erweitert werden, dass durch eine Fehlerbehandlung keine unkontrollierten Fehler mehr entstehen können.

In [23]:
from fractions import Fraction

try:
    # Eingabe des ersten Bruchs
    zaehler1 = int(input("Gib den Zähler des ersten Bruchs ein: "))
    nenner1 = int(input("Gib den Nenner des ersten Bruchs ein: "))
    
    # Eingabe des zweiten Bruchs
    zaehler2 = int(input("Gib den Zähler des zweiten Bruchs ein: "))
    nenner2 = int(input("Gib den Nenner des zweiten Bruchs ein: "))
    
    # Erstellen der Brüche
    bruch1 = Fraction(zaehler1, nenner1)
    bruch2 = Fraction(zaehler2, nenner2)
    
    # Berechnungen
    print(f"{bruch1} + {bruch2} = {bruch1 + bruch2}")
    print(f"{bruch1} - {bruch2} = {bruch1 - bruch2}")
    print(f"{bruch1} * {bruch2} = {bruch1 * bruch2}")
    print(f"{bruch1} / {bruch2} = {bruch1 / bruch2}")

except ValueError:
    print("Du hast keine gültige ganze Zahl eingegeben.")
except ZeroDivisionError as zde:
    print("Division durch 0 ist nicht definiert. Fehlermeldung:", zde)
except Exception as e:
    print(f"Ein unerwarteter Fehler ist aufgetreten: {e}")
finally:
    print("Das Programm wurde beendet.")

1/5 + 2/3 = 13/15
1/5 - 2/3 = -7/15
1/5 * 2/3 = 2/15
1/5 / 2/3 = 3/10
Das Programm wurde beendet.


## Aufgabe 2 - Fehlerbehandlung (Laufzeitfehler)
Das folgende Programm soll so erweitert werden, dass mögliche Laufzeitfehler abgefangen werden. Dabei soll die Fallunterscheidung mit Hilfe der Diskriminante wegfallen.

In [26]:
# für die Berechnung der Quadratwurzel benötigen wir das Modul "math"
import math as math

a = float(input("Gib die den Koeffizienten a ein:"))
b = float(input("Gib die den Koeffizienten b ein:"))
c = float(input("Gib die den Koeffizienten c ein:"))

diskriminante = b ** 2 - 4 * a * c

if diskriminante >= 0:
    if diskriminante == 0:
        loesung = - b / (2 * a)
        print("Die Doppellösung lautet:", loesung)
    else:
        loesung_1 = (- b + math.sqrt(diskriminante)) / (2 * a)
        loesung_2 = (- b - math.sqrt(diskriminante)) / (2 * a)
        print("Die beiden Lösungen lauten:")
        print("x1=", loesung_1)
        print("x2=", loesung_2)
else:
    print("Die Gleichung besitzt keine reellen Lösungen!")

Die Gleichung besitzt keine reellen Lösungen!


In [30]:
import math

try:
    a = float(input("Gib den Koeffizienten a ein: "))
    b = float(input("Gib den Koeffizienten b ein: "))
    c = float(input("Gib den Koeffizienten c ein: "))
   
    # Berechnung der Diskriminante
    diskriminante = b ** 2 - 4 * a * c
    
    # Berechnung der Lösungen, ohne die Diskriminante zu überprüfen
    loesung_1 = (-b + math.sqrt(diskriminante)) / (2 * a)
    loesung_2 = (-b - math.sqrt(diskriminante)) / (2 * a)
    
    print("Die beiden Lösungen lauten:")
    print("x1 =", loesung_1)
    print("x2 =", loesung_2)

except ValueError as ve:
    print(f"Ungültige Eingabe: {ve}")
except ZeroDivisionError:
    print("Division durch Null aufgetreten. Der Koeffizient a darf nicht 0 sein.")
except Exception as e:
    print(f"Ein unerwarteter Fehler ist aufgetreten: {e}")
finally:
    print("Das Programm wurde beendet.")

Ungültige Eingabe: math domain error
Das Programm wurde beendet.


## Aufgabe 3 - Fehlerbehandlung (Debugging)
### Türstatus mit Gültigkeitsprüfung
Ein Türsensor gibt den Status einer Tür an ("offen" oder "geschlossen"). Schreibe ein Programm, das den aktuellen Status überprüft und eine Meldung ausgibt. Das Programm soll dich nach einem möglich Status befragen und die Eingabe überprüfen, ob ein gültiger Status eingegeben wurde.

Setze in Zeile 2 einen _Haltepunkt_ und debugge diese Zelle und lass dir die entsprechenden Variableninhalte im Einzelschritt anzeigen, so dass du den Programmfluss nachvollziehen kannst. Das heisst, definiere die folgenden Ausdrücke zum überwachen:
-   gueltiger_status
-   gueltig
-   tuer_status.

Vegiss nicht, den Haltepunkt wieder zu löschen.   


In [31]:
# Gültige Stati festlegen
gueltiger_status = ['offen', 'geschlossen']

gueltig = False
while not(gueltig):
    tuer_status = input('Gib einen Türstatus ein: "offen" / "geschlossen":')
    if tuer_status in gueltiger_status:
        gueltig = True
        
# Selektion zur Überprüfung des Türstatus
if tuer_status == 'offen':
    print('Achtung: Die Tür ist offen!')
else:
    print('Die Tür ist geschlossen.')    

Achtung: Die Tür ist offen!


## Aufgabe 4 - Fehlerbehandlung (Debugging)
Das folgende Python-Programm, generiert 7 Zufallszahlen im Intervall I =[0; 10]. Erweitere das Programm so, dass du es zeilenweise debuggen kannst. Überwache die folgenden Ausdrücke oder Variablen:
-   zufalls_zahlen
-   i
-   zufalls_zahl.

Setze einen _Haltepunkt_ in der Zeile 4.

In [1]:
from random import randint

# festgelegte Anzahl Zufallszahlen
anzahl = 7

zufalls_zahlen = []
for i in range(anzahl):
    zufalls_zahl = randint(0, 10)
    zufalls_zahlen.append(zufalls_zahl)

print(zufalls_zahlen)

[9, 6, 6, 8, 0, 9, 3]


## Aufgabe 5 - Daten aus Datei lesen und in Datei schreiben
Häufig sind Daten in Dateien gespeichert und müssen zur Verarbeitung, Analyse, Suche, usw. in den Speicher eingelesen werden. Dazu gibt es in Python zahlreiche Möglichkeiten. Dies hängt auch davon ab, in welchen Dateitypen die Daten gespeichert sind.<br>
Es ist nicht nötig dafür die Programmbibliothek __pandas__ zu importieren, aber sie stellt weitere Funktionen zur Verfügung, welche zur Datenanalyse eingesetzt werden. Du kannst diese mächtige Bibliothek nach eigenem Interesse erkunden.<br>
- [Pandas - Wiki](https://de.wikipedia.org/wiki/Pandas_(Software))
- [Pandas - pypi](https://pypi.org/project/pandas/)

#### a) Daten aus Datei einlesen
Erstelle ein Python-Programm, mit welchem du die Datei "daten-ein-8-5.txt" einlesen kannst. Verwende dazu die Bibliothek [pandas](https://pandas.pydata.org/docs/index.html#). In dieser Datei sind in einer Spalte numerische Werte einer produzierenden Maschine gespeichert. Die Spalte verfügt über keine Überschrift.

In [32]:
import pandas as pd

# a) Einlesen der Daten mit zwei wichtigen Parametern: 
# 1) decimal legt fest, wie das Dezimalkomma definiert ist
# 2) header legt fest, ob in den Daten eine Spaltenbeschriftung existiert oder nicht

# allenfalls musst du das Datenverzeichnis anpassen. 
datei_eingabe       = "Daten/daten-ein-8-5.txt"
datei_ausgabe       = "Daten/daten-aus-8-5.txt"
datei_ausgabe_excel = "Daten/daten-aus-8-5-excel.xlsx"

# Einlesen der Daten mit Hilfe der importierten Bibliothek pandas,
# auf welche wir wegen des obigen Importbefehls mit pd verwenden können.
daten = pd.read_csv(datei_eingabe, decimal ='.', header = None)

print(daten)

             0
0    23.042313
1    27.098242
2    25.166187
3    23.473216
4    24.240603
..         ...
995  24.107926
996  23.771814
997  27.345228
998  26.759573
999  26.032696

[1000 rows x 1 columns]


#### b) erste 5 Zeilen anzeigen
Erweitere das Programm so, dass dir die ersten 5 Zeilen der Daten angezeigt werden. Dies kannst du durch den Befehl `variable.head()` realisieren.

In [33]:
# nur die ersten 5 Zeilen ausgeben
print(daten.head())

           0
0  23.042313
1  27.098242
2  25.166187
3  23.473216
4  24.240603


#### c) Wichtigste Parameter anzeigen
Mit der Methode `variable.describe()` kannst du dir die wichtigsten Parameter der Daten ausgeben lassen. Welche sind das?

Siehe dazu auch: [Quartile - Wiki](https://de.wikipedia.org/wiki/Empirisches_Quantil#Spezielle_Quantile) oder in [Quartile - Studyflix](https://studyflix.de/statistik/quartil-1628) und [Standardabweichung - Studyflix](https://studyflix.de/statistik/standardabweichung-1042)

-   _count_: liefert die Anzahl Zahlen
-   _mean_: liefert das arithmetische Mittel der Zahlen
-   _std_: liefert die Standardabweichung
-   _min_: die kleinste Zahl
-   _25%_: unteres Quartil
-   _50%_: mittleres Quartil (Median)
-   _75%_: oberes Quartil
-   _max_: die grösste Zahl


In [34]:
# die wichtigsten Parameter ausgeben
print(daten.describe())

                 0
count  1000.000000
mean     25.413497
std       2.493278
min      18.309980
25%      23.734005
50%      25.478610
75%      27.042163
max      33.683508


#### d) Daten in Datei ausgeben
Jetzt soll eine Spaltenüberschrift mit der Methode `variable.columns = ['Zahlen']` hinzugefügt und die Datei in die Datei _daten-aus-8-5.txt_ gespeichert werden.

In [35]:
# Spaltenüberschrift hinzufügen
daten.columns = ['Zahlen']

# die Daten, mit der ergänzten Spalte werden in die folgende Datei geschrieben.
# Der Parameter "index=0" bezweckt, dass keine Index-Spalte geschrieben wird. Wenn dieser
# weggelassen wird, dann wird eine Spalte eingefügt, welche die einzelnen Zeilen nummeriert.
# Bitte selber ausprobieren...
daten.to_csv(datei_ausgabe, decimal='.', header=True, index=0)

#### e) Daten in Excel-Datei ausgeben
Die gleiche Datei soll jetzt als Excel-Datei gespeichert werden. Es kann sein, dass du hierzu noch folgende Module installieren musst:<br>

- Modul "xlwt" - zum Speichern von "xls-Dateien"
- Modul "openpyxl" - zum Speichern von "xlsx-Dateien"

Damit du dir die Daten (Excel-Datei) direkt in VS-Code anschauen kannst, musst du im VS Code die Erweiterung "Office Viewer" installieren.

In [36]:
# die Daten, mit der ergänzten Spalte werden in die folgende Datei geschrieben:
# als "xlsx-Datei"
daten.to_excel(datei_ausgabe_excel, index=0)

## Aufgabe 6 - Zufallszahlen in eine Datei schreiben (mit vorgegebenen Parametern)
Häufig braucht man Daten, um gewisse Auswertungen kontrollieren zu können. Deshalb soll hier gezeigt werden, wie solche generiert werden können. Die Vorgaben sind oft die Lage- und Streuparameter wie: Mittelwert, Standardabweichung, Verteilung und Anzahl der Daten.<br> 
Die vorliegenden Daten werden von einer Abfüllmaschine ausgelesen. Diese füllt eine vorgegebene Menge (100 g) an Schokolade ab. Die internen Qualitätsvorgaben sollen überprüft werden. Dabei wird festgelegt, dass die Abfüllmenge nicht mehr als 3 g von diesen 100 g abweichen darf.

#### a) Zufallszahlen nach Vorgabe generieren
Erstelle ein Python-Programm, mit welchem eine Anzahl (n = 1500) vorgegebener Zufallszahlen mit bestimmtem Mittelwert und Streueung (Standardabweichung) berechnet werden. Diese sollen auf eine Nachkommastelle gerundet in einer Datei _daten-aus-8-6.txt_ gespeichert werden.

In [37]:
import numpy as np
import pandas as pd

# die Zufallsdaten der Maschinenproduktion
datei_ausgabe         = "Daten/daten-aus-8-6.txt"
# diejenigen Daten, welche die Vorgabe nicht erfüllen (zu wenig abgefüllt)
datei_ausgabe_weniger = "Daten/daten-aus-8-6-weniger.txt"
# diejenigen Daten, welche die Vorgabe nicht erfüllen (zu viel abgefüllt)
datei_ausgabe_mehr    = "Daten/daten-aus-8-6-mehr.txt"

mittelw  = 100      # der Mittelwert (Sollwert) der zufällig generierten Zahlen wird festgelegt
standabw = 3        # die Streuung wird hier auf drei festgelegt
n        = 1500     # die Anzahl der Zufallszahlen wird definiert

# bestimmt die Zufallszahlen mit den gewünschten Parametern/Verteilung in einem Array
np.random.seed(1)
daten    = np.random.normal(mittelw, standabw, n)
gerundet = np.around(daten, decimals = 1)

# die gerundeten Zufallszahlen werden in eine Textdatei gespeichert
# np.savetxt(datei_ausgabe, gerundet, fmt = '%2.1f')
gerundet = pd.DataFrame(gerundet)
gerundet.to_csv(datei_ausgabe, decimal='.', header=False, index=0)


#### b) Daten einlesen und Spaltenüberschrift ergänzen
Erstelle ein Python-Programm, mit welchem die in a) generierten und gespeicherten _daten-aus-8-6.txt_ Zufallszahlen eingelesen werden können. Es soll eine Spaltenüberschrift mit der Methode `variable.columns = ['Masse']` hinzugefügt werden.

In [38]:
import pandas as pd

datei_eingabe = "Daten/daten-aus-8-6.txt"

# Die generierten Daten werden eingelsen
daten = pd.read_csv(datei_eingabe, decimal ='.', header = None)
# wir ergänzen noch eine Spaltenüberschrift
daten.columns = ['Masse']

#### c) Wichtigste Parameter ausgeben
Erweitere dein Programm und gib die wichtigsten Parameter (beschreibende Statistik) mit der Methode `variable.describe()` aus. Vergleiche diese Werte mit den zuvor vorgegebenen Parametern bei der Generierung.

In [39]:
# Übersicht mit den wichtigsten Parametern
print(daten.describe())

             Masse
count  1500.000000
mean    100.157200
std       2.987045
min      90.500000
25%      98.200000
50%     100.200000
75%     102.200000
max     111.900000


#### d) Abweichende Datensätze
Jetzt soll dein Program die Anzahl abweichenden Datensätze berechnen und ausgeben. Also diejenigen, welche mehr oder weniger als 3 g vom Mittelwert abweichen. Diese erfüllen die Qualitätsvorgabe nicht. Dazu kannst du eine Abfrage (Filter) anwenden. Dies erreichst du mit der Methode `variable.query('spaltenüberschrift < zahl')`  oder `variable.query('spaltenüberschrift > zahl')`.

In [44]:
# zu schwere oder zu leichte Verpackungen bestimmen/filtern
weniger = daten.query('Masse < 97.0')
mehr    = daten.query('Masse > 103.0')

print("Es gibt", len(weniger), "Verpackungen, deren Masse weniger beträgt als 97")
print("Es gibt", len(mehr),    "Verpackungen, deren Masse mehr als 103")

Es gibt 206 Verpackungen, deren Masse weniger beträgt als 97
Es gibt 241 Verpackungen, deren Masse mehr als 103


#### e) Ausgabe in Textdateien
Jetzt speichern wir die Daten in Textdateien auf einem externen Speichermedium (Harddisk):

In [45]:
datei_ausgabe_weniger = "Daten/daten-aus-8-6-weniger.txt"
datei_ausgabe_mehr = "Daten/daten-aus-8-6-mehr.txt"

# Ausgabe der Datensätze, welche mehr Inhalt aufweisen, als toleriert
# dabei wird auch die ursprüngliche Zeilennummer (Index) gespeichert
mehr = mehr.rename_axis('Index').reset_index()
mehr.to_csv(datei_ausgabe_mehr, sep=';', decimal='.', header=True, index = False)

# Ausgabe der Datensätze, welche weniger Inhalt aufweisen, als toleriert
# dabei wird auch die ursprüngliche Zeilennummer (Index) gespeichert
weniger = weniger.rename_axis('Index').reset_index()
weniger.to_csv(datei_ausgabe_weniger, sep=';', decimal='.', header=True, index = False)


#### f) Interpretation der Ergebnisse
Interpretiere die Ergebnisse aus Sicht des Kunden, der die Schokoladen-Päckchen einkauft und aus der Sicht der Produktion.


<span style="color: green;">
Die 206 Packungen, welche weniger als 97 g Schokolade enthalten, sind aus der Sicht des Kunden ein Verlust, da er mehr bezahlt hat, als er erhält. Hingegen weisen 241 Packungen mehr als 103 g aus, was zu Gunsten des Kunden ist. Die Sicht der Firma ist entsprechend umgekehrt.
</span>

#### g) Verlust oder Gewinn?
Wie würdest du vorgehen, wenn du weisst, dass 1 g Schokolade 0.80 CHF kostet, um auszurechnen, wieviel Verlust oder Gewinn die Firma mit dieser Produktionsserie ausweist? Brauchst du noch weitere Daten?

In [47]:
import pandas as pd

datei_mehr      = "Daten/daten-aus-8-6-mehr.txt"
datei_weniger   = "Daten/daten-aus-8-6-weniger.txt"

untere_grenze =  97
obere_grenze  = 103
preis_gramm   = 0.8

# Die generierten Daten werden eingelesen
mehr    = pd.read_csv(datei_mehr, decimal ='.', sep=';')
weniger = pd.read_csv(datei_weniger, decimal ='.', sep=';')

# wir berechnen die jeweilige Summe
summe_mehr    = mehr['Masse'].sum()
summe_weniger = weniger['Masse'].sum()

# wir müssen jetzt noch je Zwischenresultat die erlaubte Masse
# je Packung subtrahieren, so dass wir die effektiven Differenzen
# vorliegen haben.
summe_mehr_eff = summe_mehr - (len(mehr) * obere_grenze)
summe_weniger_eff = (len(weniger) * untere_grenze) - summe_weniger

resultat_mehr = summe_mehr_eff * preis_gramm
resultat_weniger = summe_weniger_eff * preis_gramm

print(f"Zu viel produziert in CHF: {resultat_mehr:.2f}")
print(f"Zu wenig produziert in CHF: {resultat_weniger:.2f}", )

Zu viel produziert in CHF: 314.32
Zu wenig produziert in CHF: 282.88


<span style="color: green;">
Unter dem Strich hat die Firma ca. 31.44 CHF Verlust. Allenfalls müssten wir weitere Daten zur effektiven Kostenrechnung hinzufügen. Beispielsweise das entsprechende Zeitintervall, in welchem die Stichprobe entnommen worden ist. Grösse der Grundgesamtheit o.ä.
</span>

## Aufgabe 7 - Wohnungstemperaturen und Heizölverbrauch

#### a) Daten generieren
Erstelle ein Pythonprogramm, mit welchem du die Temperaturen und den Ölverbrauch (durchschnittlich pro Tag) für eine Wohnung im Gebäude nach den Vorgaben:
-   Temperatur:
    -   Mittelwert = 20 Grad Celsius
    -   Standardabweichung (Streuung) = 2 Grad Celsius
    -   Anzahl Werte n = 365 (1 Jahr)

-   Ölverbrauch:
    -   Mittelwert = 8 l
    -   Standardabweichung (Streuung) = 2 Grad Celsius
    -   Anzahl Werte n = 365 (1 Jahr)

zufällig und ohne Abhängigkeit von der Jahreszeit und unabhängig voneinander generieren kannst. Speichere die Daten in einer Datei mit dem Namen _daten_aus_8_7.txt_ ab.

In [48]:
import numpy as np
import pandas as pd

# die Zufallsdaten der Wohnungstemperaturen und des Heizölverbrauches
datei_ausgabe = "Daten/daten-aus-8-7.txt"

mittelw_oel   =   8     # der Mittelwert (Sollwert) der zufällig generierten Zahlen wird festgelegt
standabw_oel  =   1     # die Streuung 

mittelw_temp  =  20     # der Mittelwert (Sollwert) der zufällig generierten Zahlen wird festgelegt
standabw_temp =   2     # die Streuung 
n             = 365     # die Anzahl der Zufallszahlen wird definiert

# bestimmt die Zufallszahlen mit den gewünschten Parametern/Verteilung in einem Array
np.random.seed(1)
daten_oel = np.random.normal(mittelw_oel, standabw_oel, n)
daten_temp = np.random.normal(mittelw_temp, standabw_temp, n)
# rundet auf eine Nachkommastelle
gerundet_oel = np.around(daten_oel, decimals=1)
gerundet_temp = np.around(daten_temp, decimals=1)

# Erstellen eines DataFrames mit beiden Datensätzen
gerundet = pd.DataFrame({
    'Oelverbrauch': gerundet_oel,
    'Temperatur': gerundet_temp
})

# die gerundeten Zufallszahlen werden in eine Textdatei gespeichert
gerundet.to_csv(datei_ausgabe, decimal='.', header=True, index=False)

#### b) Daten einlesen und auswerten (beschreibende Statistik)
Lies die soeben generierten Daten aus der Datei _daten-aus-8-7.txt_ ein und gib:
-   die ersten 5 Zeilen aus
-   gib die wichtigsten Parameter je Temperaturen und Ölverbrauch aus.

In [49]:
import pandas as pd

# Datei einlesen
datei_ausgabe = "Daten/daten-aus-8-7.txt"
daten = pd.read_csv(datei_ausgabe)

# Erste fünf Zeilen ausgeben
print("Erste fünf Zeilen der Daten:")
print(daten.head())

# Wichtige Parameter mit describe() angeben
print("\nWichtige Parameter für Ölverbrauch und Temperatur:")
print(daten.describe())

Erste fünf Zeilen der Daten:
   Oelverbrauch  Temperatur
0           9.6        21.5
1           7.4        21.0
2           7.5        14.4
3           6.9        21.2
4           8.9        20.6

Wichtige Parameter für Ölverbrauch und Temperatur:
       Oelverbrauch  Temperatur
count    365.000000  365.000000
mean       8.053699   20.061370
std        0.961323    2.045905
min        5.200000   14.400000
25%        7.400000   18.800000
50%        8.000000   20.000000
75%        8.700000   21.500000
max       11.000000   27.900000
